app/src/main/java/tech/ula/utils/AndroidUtility.kt

```diff
@@ -280,24 +267,16 @@ class DownloadManagerWrapper {
         return ""
     }
 
-    fun getDownloadsDirectory(): File {
-        return Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS)
-    }
-}
-
-class DownloadBroadcastReceiver : BroadcastReceiver() {
-    private var doOnReceived: (Long) -> Unit = {}
-
-    fun setDoOnReceived(action: (Long) -> Unit) {
-        doOnReceived = action
+    fun downloadHasNotFailed(cursor: Cursor): Boolean {
+        if (cursor.moveToFirst()) {
+            val status = cursor.getInt(cursor.getColumnIndex(DownloadManager.COLUMN_STATUS))
+            return status != DownloadManager.STATUS_FAILED
```

Comment:
I see in `MainActivityViewModel` line 228 that when the state is set to `DownloadsHaveFailed`, that the message the `IllegalState` reason is set to "Failed to download assets."  I think this could be more specific.  When the `DownloadManager`'s status is set to `STATUS_FAILED`, the  `COLUMN_REASON` is updated.  We could use this for the `IllegalState` reason and it may help us better debug why the downloads failed.

[android docs on DownloadManager's COLUMN_REASON property](https://developer.android.com/reference/kotlin/android/app/DownloadManager#COLUMN_REASON%3Akotlin.String)